marconi/src/Marconi/Index/Utxo.hs

```diff
@@ -1,185 +1,213 @@
 import Database.SQLite.Simple.FromRow (FromRow (fromRow), field)
 import Database.SQLite.Simple.ToField (ToField (toField))
 import Database.SQLite.Simple.ToRow (ToRow (toRow))
 import GHC.Generics (Generic)
 import System.Random.MWC (createSystemRandom, uniformR)
-import Text.ParserCombinators.Parsec (parse)
 import Text.RawString.QQ (r)
 
+import Cardano.Api ()
 import Cardano.Api qualified as C
 import "cardano-api" Cardano.Api.Shelley qualified as Shelley
-import Cardano.Binary (fromCBOR, toCBOR)
 import Marconi.Types (CurrentEra, TargetAddresses, TxOut, pattern CurrentEra)
-import RewindableIndex.Index.VSqlite (SqliteIndex)
-import RewindableIndex.Index.VSqlite qualified as Ix
-
-data Utxo                       = Utxo
-    { _utxoAddress          :: !C.AddressAny
-    , _utxoTxId             :: !C.TxId
-    , _utxoTxIx             :: !C.TxIx
-    , _utxoDatum            :: Maybe C.ScriptData
-    , _utxoDatumHash        :: Maybe (C.Hash C.ScriptData)
-    , _utxoValue            :: C.Value
-    , _utxoInlineScript     :: Maybe ByteString
-    , _utxoInlineScriptHash :: Maybe (C.ScriptHash)
-    } deriving (Show, Generic)
+import RewindableIndex.Storable (Buffered (getStoredEvents, persistToStorage), HasPoint (getPoint),
+                                 QueryInterval (QEverything, QInterval), Queryable (queryStorage),
+                                 Resumable (resumeFromStorage), Rewindable (rewindStorage), StorableEvent,
+                                 StorableMonad, StorablePoint, StorableQuery, StorableResult, emptyState,
+                                 filterWithQueryInterval)
+import RewindableIndex.Storable qualified as Storable
 
-$(makeLenses ''Utxo)
 
-instance Eq Utxo where
-    u1 == u2 = (_utxoTxId u1) == (_utxoTxId u2)
 
-instance Ord Utxo where
-    compare u1 u2 = compare (_utxoTxId u1) (_utxoTxId u2)
+type UtxoIndexer = Storable.State UtxoHandle
 
-data UtxoEvent          = UtxoEvent
-    { _utxoEventUtxos   :: [Utxo]
-    , _utxoEventInputs  :: !(Set C.TxIn)
-    , _utxoEventSlotNo  :: !C.SlotNo
-    , _utxoEventBlockNo :: !C.BlockNo
-    } deriving (Show, Eq)
+data UtxoHandle = UtxoHandle
+  { hdlConnection :: SQL.Connection
+  , hdlDpeth      :: Int
+  }
 
-$(makeLenses ''UtxoEvent)
+newtype instance StorableQuery UtxoHandle =
+  UtxoAddress C.AddressAny deriving (Show, Eq)
 
-data UtxoRow            = UtxoRow
-    { _utxoRowUtxo    :: Utxo
-    , _utxoRowSlotNo  :: !C.SlotNo
-    , _utxoRowBlockNo :: !C.BlockNo
-    } deriving (Show, Eq, Ord, Generic)
+type QueryableAddresses = NonEmpty (StorableQuery UtxoHandle)
 
-$(makeLenses ''UtxoRow)
+type instance StorableMonad UtxoHandle = IO
 
-type Result = Maybe [UtxoRow]
+type instance StorablePoint UtxoHandle = C.ChainPoint
 
-type Notification = ()
+newtype Depth = Depth Int
 
-type UtxoIndex
-    = SqliteIndex
-      UtxoEvent
-      Notification
-      C.AddressAny
-      Result
+data Utxo = Utxo
+  { _address          :: StorableQuery UtxoHandle
```

Comment:
Why not just have the field `C.AddressAny`? Seems circular to include indexer framework's "main types" within an event.